File: src/main/storage/repos/metadata-cache.ts
import { and, desc, eq, notInArray } from 'drizzle-orm';
import { log } from '../../log';
import type { KiraDb } from '../db';
import { metadataCache } from '../schema/metadata-cache';

export type MetaKind = 'children' | 'describe' | 'definition';

// The unique index is (connection_id, path) — `kind` is not part of the key, so a `children`
// payload and a `describe` payload for the same path share one row. Both live under this shape
// in `payload_json`; `kind` on the row is informational (whichever was written last).
interface CachedPayload {
  children?: unknown;
  describe?: unknown;
  definition?: unknown;
}

const MAX_PAYLOAD_BYTES = 4 * 1024 * 1024;
// P43 iter2 F15/D20: a Browse session writes one row per level navigated, each up to
// MAX_PAYLOAD_BYTES, with no eviction until the connection's next successful connect
// (main/connections.ts's own dropCached) — a single session that browses deep leaves the on-disk
// cache unbounded until then. Per-connection (not global) so one heavily-browsed S3 connection
// cannot evict a small Postgres connection's whole tree. Same order as db.ts's own
// STMT_CACHE_MAX and repos/ops.ts's HARD_CAP_ROWS; an evicted level costs one round trip to
// re-fetch.
const MAX_ROWS_PER_CONNECTION = 200;

async function readPayload(
  db: KiraDb,
  connectionId: string,
  path: string,
): Promise<CachedPayload | null> {
  const rows = await db
    .select({ payloadJson: metadataCache.payloadJson })
    .from(metadataCache)
    .where(and(eq(metadataCache.connectionId, connectionId), eq(metadataCache.path, path)));
  const row = rows[0];
  if (!row) return null;
  try {
    return JSON.parse(row.payloadJson) as CachedPayload;
  } catch {
    return null;
  }
}

// JSON.parse'd, NOT validated here — callers parse through the domain Zod schema and drop the
// row (dropCached) on a shape mismatch, treating it as a miss rather than surfacing an error.
export async function getCached(
  db: KiraDb,
  connectionId: string,
  path: string,
  kind: MetaKind,
): Promise<unknown | null> {
  const payload = await readPayload(db, connectionId, path);
  return payload?.[kind] ?? null;
}

export async function putCached(
  db: KiraDb,
  connectionId: string,
  path: string,
  kind: MetaKind,
  payload: unknown,
): Promise<void> {
  await db.transaction(async (tx) => {
    const rows = await tx
      .select({ payloadJson: metadataCache.payloadJson })
      .from(metadataCache)
      .where(and(eq(metadataCache.connectionId, connectionId), eq(metadataCache.path, path)));
    let existing: CachedPayload = {};
    if (rows[0]) {
      try {
        existing = JSON.parse(rows[0].payloadJson) as CachedPayload;
      } catch {
        existing = {};
      }
    }
    const merged: CachedPayload = { ...existing, [kind]: payload };
    const json = JSON.stringify(merged);
    if (Buffer.byteLength(json, 'utf8') > MAX_PAYLOAD_BYTES) {
      log(
        'warn',
        'storage/metadata-cache',
        `payload for ${connectionId}:${path} exceeds 4 MB, not cached`,
      );
      return;
    }
    const fetchedAt = new Date().toISOString();
    await tx
      .insert(metadataCache)
      .values({ connectionId, path, kind, payloadJson: json, fetchedAt, etag: null })
      .onConflictDoUpdate({
        target: [metadataCache.connectionId, metadataCache.path],
        set: { kind, payloadJson: json, fetchedAt },
      });

    // D20's eviction pass: keep this connection's MAX_ROWS_PER_CONNECTION newest-`fetched_at`
    // rows, drop the rest. An `onConflictDoUpdate` above never grows the connection's row count,
    // so this only ever has anything to do right after the insert branch actually added a row —
    // but running it unconditionally is one cheap indexed query against a table this small, not
    // a cost worth special-casing around.
    const keep = await tx
      .select({ path: metadataCache.path })
      .from(metadataCache)
      .where(eq(metadataCache.connectionId, connectionId))
      .orderBy(desc(metadataCache.fetchedAt))
      .limit(MAX_ROWS_PER_CONNECTION);
    const keepPaths = keep.map((r) => r.path);
    if (keepPaths.length > 0) {
      await tx
        .delete(metadataCache)
        .where(
          and(
            eq(metadataCache.connectionId, connectionId),
            notInArray(metadataCache.path, keepPaths),
          ),
        );
    }
  });
}

// `path` omitted = drop every cached row for the whole connection.
export async function dropCached(db: KiraDb, connectionId: string, path?: string): Promise<void> {
  if (path === undefined) {
    await db.delete(metadataCache).where(eq(metadataCache.connectionId, connectionId));
  } else {
    await db
      .delete(metadataCache)
      .where(and(eq(metadataCache.connectionId, connectionId), eq(metadataCache.path, path)));
  }
}

File: apps/kira-studio/internal/storage/repos/metadata_cache.go
package repos

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"log/slog"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/storage/model"
)

// maxMetadataPayloadBytes and maxMetadataRowsPerConnection mirror metadata-cache.ts's
// MAX_PAYLOAD_BYTES and MAX_ROWS_PER_CONNECTION (P43 iter2 F15/D20).
const (
	maxMetadataPayloadBytes      = 4 * 1024 * 1024
	maxMetadataRowsPerConnection = 200
)

type MetadataCacheRepo struct {
	DB *sql.DB
}

// readPayload returns the row's payload object (or nil if there is no row / it fails to parse —
// treated as a miss, mirroring metadata-cache.ts's own try/catch).
func readPayload(q interface {
	QueryRow(query string, args ...any) *sql.Row
}, connectionID, path string) map[string]json.RawMessage {
	var payloadJSON string
	err := q.QueryRow(
		`SELECT payload_json FROM metadata_cache WHERE connection_id = ? AND path = ?`,
		connectionID, path,
	).Scan(&payloadJSON)
	if err != nil {
		return nil
	}
	var payload map[string]json.RawMessage
	if err := json.Unmarshal([]byte(payloadJSON), &payload); err != nil {
		return nil
	}
	return payload
}

// fetchedAtKey is the reserved payload key holding each kind's own last-write timestamp (P24 D3):
// { "children": …, "columns": …, "fetchedAt": {"children": "…", "columns": "…"} }. The four kind
// names are a closed set fixed in internal/tree/service.go, so this can never collide with a
// payload key.
const fetchedAtKey = "fetchedAt"

// fetchedAtFor reads kind's own entry out of payload's reserved fetchedAt map — "" when the
// map or the entry is absent, which is the correct answer for both a pre-P24 row (D2's upgrade
// path) and a row that has simply never cached this kind.
func fetchedAtFor(payload map[string]json.RawMessage, kind string) string {
	raw, ok := payload[fetchedAtKey]
	if !ok {
		return ""
	}
	var m map[string]string
	if err := json.Unmarshal(raw, &m); err != nil {
		return ""
	}
	return m[kind]
}

// Get is JSON.parse'd, NOT further validated here — callers parse through their own domain
// shape and treat a mismatch as a miss (metadata-cache.ts's own discipline). The second return is
// the payload's own per-kind fetch time (P24 D3) — internal/tree.Service is the sole judge of
// whether that makes it stale.
func (r *MetadataCacheRepo) Get(connectionID, path, kind string) (json.RawMessage, string, error) {
	payload := readPayload(r.DB, connectionID, path)
	return payload[kind], fetchedAtFor(payload, kind), nil
}

// Put merges payload into the existing row's {children?, describe?, definition?, columns?} object
// (the unique index is (connection_id, path) — kind is not part of the key, so a 'children'
// payload and a 'describe' payload for the same path share one row) and stamps this kind's own
// entry in the reserved fetchedAt map (P24 D3) — every other kind's payload AND fetch time is left
// untouched. Then runs D20's per-connection eviction — all in one transaction.
func (r *MetadataCacheRepo) Put(connectionID, path, kind string, payload json.RawMessage) error {
	tx, err := r.DB.Begin()
	if err != nil {
		return fmt.Errorf("repos/metadata_cache: begin: %w", err)
	}
	defer tx.Rollback() //nolint:errcheck

	existing := readPayload(tx, connectionID, path)
	merged := map[string]json.RawMessage{}
	for k, v := range existing {
		merged[k] = v
	}
	merged[kind] = payload

	now := model.NowISO()
	fetchedAt := map[string]string{}
	if raw, ok := existing[fetchedAtKey]; ok {
		_ = json.Unmarshal(raw, &fetchedAt) // best-effort; a corrupt map just resets to {kind: now}
	}
	fetchedAt[kind] = now
	encodedFetchedAt, err := json.Marshal(fetchedAt)
	if err != nil {
		return fmt.Errorf("repos/metadata_cache: encode fetchedAt map: %w", err)
	}
	merged[fetchedAtKey] = encodedFetchedAt

	encoded, err := json.Marshal(merged)
	if err != nil {
		return fmt.Errorf("repos/metadata_cache: encode merged payload: %w", err)
	}
	if len(encoded) > maxMetadataPayloadBytes {
		slog.Warn("payload exceeds 4 MB, not cached", "scope", "storage/metadata-cache", "connectionId", connectionID, "path", path)
		return nil // not an error; the deferred Rollback is a no-op since nothing was written.
	}

	if _, err := tx.Exec(
		`INSERT INTO metadata_cache (connection_id, path, kind, payload_json, fetched_at, etag)
		 VALUES (?, ?, ?, ?, ?, NULL)
		 ON CONFLICT(connection_id, path) DO UPDATE SET kind = excluded.kind, payload_json = excluded.payload_json, fetched_at = excluded.fetched_at`,
		connectionID, path, kind, string(encoded), now,
	); err != nil {
		return fmt.Errorf("repos/metadata_cache: upsert: %w", err)
	}

	if _, err := tx.Exec(`
		DELETE FROM metadata_cache
		 WHERE connection_id = ?
		   AND path NOT IN (
		     SELECT path FROM metadata_cache
		      WHERE connection_id = ?
		      ORDER BY fetched_at DESC, rowid DESC
		      LIMIT ?
		   )
	`, connectionID, connectionID, maxMetadataRowsPerConnection); err != nil {
		return fmt.Errorf("repos/metadata_cache: evict: %w", err)
	}

	if err := tx.Commit(); err != nil {
		return fmt.Errorf("repos/metadata_cache: commit: %w", err)
	}
	return nil
}

// Drop clears one cached path (path == "" clears the whole connection — see DropConnection for
// the explicit form callers should prefer for that case).
func (r *MetadataCacheRepo) Drop(connectionID, path string) error {
	if _, err := r.DB.Exec(
		`DELETE FROM metadata_cache WHERE connection_id = ? AND path = ?`, connectionID, path,
	); err != nil {
		return fmt.Errorf("repos/metadata_cache: drop %s:%s: %w", connectionID, path, err)
	}
	return nil
}

// DropConnection drops every cached row for the whole connection (metadata-cache.ts's
// dropCached with path omitted).
func (r *MetadataCacheRepo) DropConnection(connectionID string) error {
	if _, err := r.DB.Exec(`DELETE FROM metadata_cache WHERE connection_id = ?`, connectionID); err != nil {
		return fmt.Errorf("repos/metadata_cache: drop connection %s: %w", connectionID, err)
	}
	return nil
}
